src/null-safety/migration-guide.md

```diff
@@ -1,23 +1,472 @@
+| `/*late final*/` | Mark the variable as `late final`, indicating that it has late, one-time initialization. |
+| `/*required*/`   | Mark the parameter as `required`.                                        |
+{:.table .table-striped}
+
+A single hint can have ripple effects elsewhere in the code.
+In the example from before,
+adding a `/*!*/` marker where `zero` is assigned its value (on line 2)
+makes the migration tool infer the type of `zero` as `int` instead of `int?`.
+This type change can affect code that directly or indirectly uses `zero`.
+
+```dart
+var zero = ints[0]/*!*/;
+```
+
+With the above hint, the migration tool changes its proposed edits,
+as the following code snippets show.
+Line 3 no longer has a `!` after `zero`,
+and in line 4 `zeroOne` is inferred to be
+a list of `int`, not `int?`.
+
+<table markdown="1">
+<tr>
+<th>First migration</th>
+<th>Migration with hint</th>
+</tr>
+<tr markdown="1">
+  <td markdown="1">
+```dart
+var ints = const <int?>[0, null];
+var zero = ints[0];
+var one = zero! + 1;
+var zeroOne = <int?>[zero, one];
+```
+  </td>
+  <td markdown="1">
+```dart
+var ints = const <int?>[0, null];
+var zero = ints[0]/*!*/;
+var one = zero + 1;
+var zeroOne = <int>[zero, one];
+```
+  </td>
+</tr>
+</table>
+
+
+#### Applying changes
+
+When you like all of the changes
+that the migration tool proposes, click
+**Apply migration** to tell the tool to save the migrated code
+and to update the minimum SDK in your pubspec.
+
+The next step is to [statically analyze your code](#step3-analyze).
+If it's valid, then [test your code](#step4-test).
+Then, if you've published your code on pub.dev,
+[publish a null-safe prerelease](#step5-publish).
+
+
+### Migrating by hand
+
+If you prefer not to use the migration tool,
+you can migrate manually.
+Because Dart supports [mixed-version programs][Unsound null safety],
+you can migrate one library (generally one Dart file) at a time,
+while still being able to run your app and its tests.
+
+[Unsound null safety]: /go/unsound-null-safety
+
+We recommend that you **first migrate leaf libraries** —
+libraries that don't import other files from the package.
+Then migrate libraries that directly depend on the leaf libraries.
+End by migrating the libraries that have the most
+intra-package dependencies.
+
+For example, say you have a `lib/src/util.dart` file
+that imports other (null-safe) packages and core libraries,
+but that doesn't have any `import '<local_path>'` directives.
+Consider migrating `util.dart` first,
+and then migrating simple files that depend only on `util.dart`.
+If any libraries have cyclic imports
+(for example, A imports B which imports C, and C imports A),
+consider migrating those libraries together.
+
+To migrate a package by hand, follow these steps:
+
+1. Modify the SDK constraints in the package's `pubspec.yaml` file,
+   setting the language version to 2.12 beta or later:
+   ```yaml
+environment:
+  sdk: '>=2.12.0-0 <3.0.0'
```

Comment:
Tim suggested in the blog post draft that we highlight the `-0` part, which I think is a good idea (this is not something most have seen before). I added a commit for this over on the other page; perhaps we can use something similar here?

https://github.com/dart-lang/site-www/commit/5b32909f51bd43449fc132d5aebf09e17dcea138